src/iface/ethernet.rs

```diff
@@ -644,3 +645,157 @@ impl<'a, 'b, 'c, DeviceT: Device + 'a> Interface<'a, 'b, 'c, DeviceT> {
         })
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::Packet;
+    use phy::{Loopback, ChecksumCapabilities};
+    use wire::{ArpOperation, ArpPacket, ArpRepr};
+    use wire::{EthernetAddress, EthernetFrame, EthernetProtocol};
+    use wire::{IpAddress, IpCidr, IpProtocol, IpRepr};
+    use wire::{Ipv4Address, Ipv4Repr};
+    use iface::{ArpCache, SliceArpCache, EthernetInterface};
+    use socket::SocketSet;
+
+    fn loopback_test<F>(f: F)
+            where F: FnOnce(&mut EthernetInterface<Loopback>) -> () {
+        // Create a basic device
+        let mut device = Loopback::new();
+
+        let mut arp_cache_entries: [_; 8] = Default::default();
+        let mut arp_cache = SliceArpCache::new(&mut arp_cache_entries[..]);
+
+        let mut ip_addrs = [IpCidr::new(IpAddress::v4(127, 0, 0, 1), 8)];
+        let mut iface = EthernetInterface::new(
+            &mut device, &mut arp_cache as &mut ArpCache,
+            EthernetAddress::default(), &mut ip_addrs[..], None);
+
+        f(&mut iface)
+    }
+
+    #[test]
+    fn no_icmp_to_broadcast() {
+        loopback_test(|iface| {
+            let mut eth_bytes = vec![0u8; 34];
+
+            // Unknown Ipv4 Protocol
+            //
+            // Because the destination is the broadcast address
+            // this should not trigger and Destination Unreachable
+            // response. See RFC 1122 § 3.2.2.
+            let repr = IpRepr::Ipv4(Ipv4Repr {
+                src_addr:    Ipv4Address([0x7f, 0x00, 0x00, 0x01]),
+                dst_addr:    Ipv4Address::BROADCAST,
+                protocol:    IpProtocol::Unknown(0x0c),
+                payload_len: 0,
+                ttl:         0x40
+            });
+
+            let frame = {
+                let mut frame = EthernetFrame::new(&mut eth_bytes);
+                frame.set_dst_addr(EthernetAddress::BROADCAST);
+                frame.set_src_addr(EthernetAddress([0x52, 0x54, 0x00, 0x00, 0x00, 0x00]));
+                frame.set_ethertype(EthernetProtocol::Ipv4);
+                repr.emit(frame.payload_mut(), &ChecksumCapabilities::default());
+                EthernetFrame::new(&*frame.into_inner())
```

Comment:
Kind of ugly, but inner type mutability differs without this

Reply:
That's a perfectly fine construct in smoltcp.